golden-path: add docs for integrations (#34342)
backstage/backstage: docs/golden-path/plugins/integrations/001-catalog.md

### What is the Software Catalog?

> The [Software Catalog](../../../features/software-catalog/index.md) is the

Reviewer: Thinking there's a need for a fair bit more deep links to specific docs for the different concepts introduced here

graph of typed entities that sits at the center of every Backstage instance.
Components, APIs, Resources, Systems, Users, Groups — anything an organization
wants to model and reason about — live in the catalog and are related to each

> other through a small set of well-known relations.

Reviewer: ```suggestion
other through a set of well-known relations.
```

maybe? 🤔

> other through a small set of well-known relations.

Reviewer 2: It would be helpful to provide a link to page that describes the set of relationships: Well-known Relations between Catalog Entities (docs/features/software-catalog/well-known-relations)

> Each entity is identified by its `apiVersion`, `kind`, and `metadata.name`

Reviewer: Not really true, the `apiVersion` isn't part of identifying it as such? 🤔 at least I think it's confusing to equate to the kind + namespace + name tuple, that's a more common and central concept. The `apiVersion` is tells you what the data means

Author: yeah bad claude moment - will update

backing store, for example.

**Catalog processors** transform entities as they flow through the ingestion
pipeline. They can read annotations, mutate spec fields, emit additional
entities, or attach extra relations. Use a processor when you want to react to
existing entities or annotations rather than provide a brand-new source.

**The catalog model** controls which kinds, versions, and spec types are
considered valid. Extending the model lets you introduce a new first-class
kind — `Todo`, for example — with its own schema, relations, and type guards.
Use a model extension when your data does not fit comfortably into any of the
built-in kinds.

A typical plugin uses one or two of these together: a provider plus a model
extension to introduce a new kind, or a processor that reads an annotation off
existing entities to expose plugin-specific data.

## Adding a new `backstage.io/todo` annotation

The lightest-weight integration is to attach to entities that already live in
the catalog. Adopters add an annotation to their `catalog-info.yaml`, and your
plugin reacts to it. There is no new kind, no new provider, and the source of
truth stays in the adopter's repository.

> ### Reserve the annotation name

Reviewer 2: This looks like a series of steps. The CNCF wants procedural information as a set of steps, instead of a series of paragraphs or subheadings. For example:

1. Reserve the annotation name.
2. Read the annotation in the backend.
3. Surface the annotation on the entity page.
4. Validate the annotation with a processor.
5. Wire the processor into a catalog backend module

Annotations are namespaced strings, and the `backstage.io/` prefix is
reserved for annotations defined by the project. For a plugin you ship
yourself, pick a namespace you control — for example

> `todo.backstage.io/source` — and document it. For the rest of this guide we

Reviewer: ```suggestion
`todo.acme.org/source` — and document it. For the rest of this guide we
```

Perhaps? Just to avoid confusion. Or `acme.org/todo-source`

> will pretend our plugin is shipped by the Backstage project and use

Reviewer 2: _and use_ should be _and uses_

`backstage.io/todo-source`.

Add a constant in your common package so backend, frontend, and any policies
reference the same string:

```ts
// plugins/todo-common/src/annotations.ts
export const TODO_SOURCE_ANNOTATION = 'backstage.io/todo-source';
```

### Read the annotation in the backend

Inside a route handler, look the entity up through the catalog client and
read the annotation off `metadata.annotations`:

```ts
// plugins/todo-backend/src/service/router.ts

> import { CatalogService } from '@backstage/plugin-catalog-node';

Reviewer: wonder if this needs a bit more context, maybe wrapping the example in a `createRouter` or such

import { TODO_SOURCE_ANNOTATION } from '@internal/plugin-todo-common';

router.get('/todos/by-entity/:ref', async (req, res) => {

>   const credentials = await httpAuth.credentials(req, { allow: ['user'] });

Reviewer: ```suggestion
  const credentials = await httpAuth.credentials(req);
```

Typically you don't want to do this and just go with the default policy, I think it's probably best to avoid in this example. That's unless we specifically want to list todos for a user, but feeling that would complicate the example

On the frontend, use the standard helpers to check for the annotation before
showing your card. This keeps the UI clean for entities that have not opted
in:

```tsx
// plugins/todo/src/components/EntityTodoContent.tsx
import { useEntity } from '@backstage/plugin-catalog-react';
import { TODO_SOURCE_ANNOTATION } from '@internal/plugin-todo-common';

export const isTodoAvailable = (entity: Entity) =>
  Boolean(entity.metadata.annotations?.[TODO_SOURCE_ANNOTATION]);

export const EntityTodoContent = () => {
  const { entity } = useEntity();
  if (!isTodoAvailable(entity)) {
    return <MissingAnnotationEmptyState annotation={TODO_SOURCE_ANNOTATION} />;
  }
  return <TodoListForEntity entityRef={stringifyEntityRef(entity)} />;
};
```

`MissingAnnotationEmptyState` is the conventional way to nudge adopters
towards adding the annotation, and it links straight to your plugin's docs.

### Validate the annotation with a processor

If you want to fail fast when somebody types the annotation wrong, register
a catalog processor that validates it during ingestion. Processors run as
part of the catalog pipeline, so a bad annotation surfaces as a refresh error

> on the entity rather than as a confusing empty state in the UI:

Reviewer: Thinking it's worth mentioning that this is highly optional, and most plugins skip this due to the installation complexity and instead fail within the plugin instead

  getProcessorName() {
    return 'TodoAnnotationProcessor';
  }

  async preProcessEntity(entity: Entity) {
    const value = entity.metadata.annotations?.[TODO_SOURCE_ANNOTATION];
    if (value !== undefined && !/^[a-z0-9-/]+$/.test(value)) {
      throw new Error(
        `Invalid value for annotation ${TODO_SOURCE_ANNOTATION}: ${value}`,
      );
    }
    return entity;
  }
}
```

Wire the processor into a catalog backend module:

```ts
// plugins/catalog-backend-module-todo/src/module.ts
import { createBackendModule } from '@backstage/backend-plugin-api';
import { catalogProcessingExtensionPoint } from '@backstage/plugin-catalog-node';
import { TodoAnnotationProcessor } from './TodoAnnotationProcessor';

export const catalogModuleTodoAnnotation = createBackendModule({
  pluginId: 'catalog',
  moduleId: 'todo-annotation',
  register(env) {
    env.registerInit({
      deps: { catalog: catalogProcessingExtensionPoint },
      async init({ catalog }) {
        catalog.addProcessor(new TodoAnnotationProcessor());
      },
    });
  },
});
```

Adopters install the module alongside the catalog backend, and the annotation
is now part of their ingestion pipeline.

> ## Custom TODO Entity Kind

Reviewer: I think this detracts from the golden path concept quite a bit, in that it's more of a collection of guides rather than a single one if we add this. In general we'd really want to golden path to be a guide for how to concretely build an actual full example project, so that in following the path your end result is a functional thing. It's fine to have a lot of it be explainers for code in the template, especially early on. But beyond that we'd really want the path to tell you exactly what code additions to make, how to verify that they work in the UI, etc.

Then again, maybe this is exactly what we want to build in the guide, given the search integration and so on. Bit weary that it might be too large to also introduce the kinds and so on. This is also the golden path to building plugins, and it's very rare that you are in this are as part of plugin development. I think this is probably better reserved for an app golden path.

Author: Happy to remove this, I was trying to strike a balance between things that may be useful now vs useful in the future and trying to figure out when is the right time to introduce them. I don't know how useful this really would be for end users especially early adopters.

Reviewer: I think it could be useful, but for an app golden path rather than plugin building. It's very rare that you encounter custom kinds for plugins, but definitely something I think worth introducing for apps.
